File: include/modules/Battery.c
#include <stdio.h>
#include <string.h>
#include <utils.h>

static unsigned short batindex = 0;
static char batsym[][5] = {" ", " ", " ", " ", " "};

int module_battery(char *value) {
  int capacity;
  char bat[256], status[16], batfilepath[64], *sym;
  unsigned batno = 0, batsymlen = sizeof(batsym) / sizeof(batsym[0]);

  sprintf(batfilepath, "/sys/class/power_supply/BAT%d/capacity", batno);
  if (fpscanf(batfilepath, "%d", &capacity) == -1)
    return -1;

  sprintf(batfilepath, "/sys/class/power_supply/BAT%d/status", batno++);
  if (fpscanf(batfilepath, "%s", status) == -1)
    return -1;

  sym = strcmp(status, "Charging") == 0 ? batsym[batindex]
        : strcmp(status, "Discharging") == 0
            ? batsym[capacity * batsymlen / 100]
            : batsym[batsymlen - 1];
  sprintf(bat, "%s %3d", sym, capacity);

  batindex = (batindex + 1) % batsymlen;
  return sprintf(value, "%s", bat);
}

File: logger/battery.go
package logger

import (
	"fmt"
	"os"
	. "smolprog/utils"
	"strconv"
	"time"
)

var (
	BAT_DIR = ""
	SYM     = []string{" ", " ", " ", " ", " "}
)

func init() {
	if entries, err := os.ReadDir("/sys/class/power_supply"); err == nil {
		for _, entry := range entries {
			if StartsWith(entry.Name(), "BAT") {
				BAT_DIR = "/sys/class/power_supply/" + entry.Name()
			}
		}
	}
}

type Battery struct{ sym_index int }

func (bat *Battery) Run(slot int, channel chan<- *Message) {
	IntervalRunner(bat, slot, channel)
}

func (_ *Battery) Interval() time.Duration {
	return time.Second
}

func (bat *Battery) Value() string {
	return SEP + "  " + bat.calculate() + "  "
}

func (bat *Battery) calculate() string {
	if capacity := bat.getCapacity(); capacity >= 0 {
		return fmt.Sprintf("%s %3d", bat.getStatus(capacity), capacity)
	}
	return "  ?"
}

func (_ *Battery) getCapacity() int {
	if capacity, err := FirstLineOf(BAT_DIR + "/capacity"); err == nil {
		if num, err := strconv.Atoi(capacity); err == nil {
			return num
		}
	}
	return -1
}

func (bat *Battery) getStatus(capacity int) string {
	bat.sym_index = (bat.sym_index + 1) % len(SYM)

	switch status, _ := FirstLineOf(BAT_DIR + "/status"); status {
	case "Charging":
		return SYM[bat.sym_index]
	case "Discharging":
		return SYM[capacity*len(SYM)/101]
	}
	return SYM[len(SYM)-1]
}
